pkg/apiserver/registry/system/supportbundle/rest.go

```diff
@@ -0,0 +1,378 @@
+		ovsCtlClient: client,
+		cache: &supportBundle{
+			SupportBundle: systemv1beta1.SupportBundle{
+				ObjectMeta: metav1.ObjectMeta{Name: mode},
+				Status:     systemv1beta1.SupportBundleStatusNone,
+			},
+		},
+	}
+	return Storage{
+		Mode:          mode,
+		SupportBundle: bundle,
+		Download:      &downloadREST{supportBundle: bundle},
+	}
+}
+
+// Storage contains REST resources for support bundle, includes status query and download.
+type Storage struct {
+	SupportBundle *supportBundleREST
+	Download      *downloadREST
+	Mode          string
+}
+
+var (
+	_ rest.Scoper          = &supportBundleREST{}
+	_ rest.Getter          = &supportBundleREST{}
+	_ rest.Creater         = &supportBundleREST{}
+	_ rest.GracefulDeleter = &supportBundleREST{}
+)
+
+type supportBundle struct {
+	systemv1beta1.SupportBundle
+	filepath string
+}
+
+// supportBundleREST implements REST interfaces for bundle status querying.
+type supportBundleREST struct {
+	bridge       string
+	mode         string
+	statusLocker sync.RWMutex
+	cancelFunc   context.CancelFunc
+	cache        *supportBundle
+	ovsCtlClient ovsctl.OVSCtlClient
+}
+
+// Create triggers a bundle generation progress. It only allows to create resource
+// which has the same name with the mode. It returns metav1.Status if there is any
+// error, otherwise it returns the SupportBundle.
+func (r *supportBundleREST) Create(ctx context.Context, obj runtime.Object, _ rest.ValidateObjectFunc, _ *metav1.CreateOptions) (runtime.Object, error) {
+	requestBundle := obj.(*systemv1beta1.SupportBundle)
+	if requestBundle.Name != r.mode {
+		return nil, errors.NewForbidden(systemv1beta1.ControllerInfoVersionResource.GroupResource(), requestBundle.Name, fmt.Errorf("only resource name \"%s\" is allowed", r.mode))
+	}
+	r.statusLocker.Lock()
+	defer r.statusLocker.Unlock()
+
+	if r.cancelFunc != nil {
+		r.cancelFunc()
+	}
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	r.cache = &supportBundle{
+		SupportBundle: systemv1beta1.SupportBundle{
+			ObjectMeta: metav1.ObjectMeta{Name: r.mode},
+			Status:     systemv1beta1.SupportBundleStatusCollecting,
+		},
+	}
+	r.cancelFunc = cancelFunc
+
+	go func() {
+		var err error
+		var b *supportBundle
+		if r.mode == "agent" {
+			b, err = r.collectAgent(ctx)
+		} else if r.mode == "controller" {
+			b, err = r.collectController(ctx)
+		}
+		func() {
+			r.statusLocker.Lock()
+			defer r.statusLocker.Unlock()
+			if err != nil {
+				klog.Errorf("Error when collecting supportBundle: %v", err)
+				r.cache.Status = systemv1beta1.SupportBundleStatusNone
+				return
+			}
+			select {
+			case <-ctx.Done():
+			default:
+				r.cache = b
+			}
+		}()
+		r.clean(ctx, b.filepath, bundleExpireDuration)
+	}()
+
+	return &r.cache.SupportBundle, nil
+}
+
+func (r *supportBundleREST) New() runtime.Object {
+	return &systemv1beta1.SupportBundle{}
+}
+
+// Get returns current status of the bundle. It only allows to query the resource
+// which has the same name of the mode.
```

Comment:
"It only allows querying the resource whose name is equal to the mode."

Reply:
Done.